File: d5/main.go
package main

import (
	_ "embed"
	"fmt"
	"math"
	"sync"
)

//go:embed in.txt
var input string

type Range struct {
	destStart   int
	sourceStart int
	Len         int
}

type Mapping []Range

func (m Mapping) Convert(source int) int {
	result := source
	for _, r := range m {
		if r.sourceStart <= source && source < r.sourceStart+r.Len {
			offset := source - r.sourceStart
			result = r.destStart + offset
			break
		}
	}
	return result
}

type Seeds []int

type Almanac struct {
	seeds    Seeds
	mappings []Mapping
}

func (a Almanac) SeedToLocation(seed int) int {
	current := seed
	for _, m := range a.mappings {
		current = m.Convert(current)
	}
	return current
}

func main() {
	alm := AlmanacFromString(input)
	part1 := math.MaxInt
	for _, seed := range alm.seeds {
		location := alm.SeedToLocation(seed)
		if location < part1 {
			part1 = location
		}
	}
	fmt.Println(part1)

	var results = make(chan int, 1000)
	wg := sync.WaitGroup{}
	for i := 0; i < len(alm.seeds); i += 2 {
		wg.Add(1)
		i := i
		go func() {
			defer wg.Done()
			curMin := math.MaxInt
			start := alm.seeds[i]
			count := alm.seeds[i+1]
			for j := start; j < start+count; j++ {
				location := alm.SeedToLocation(j)
				if location < curMin {
					curMin = location
				}
			}
			results <- curMin
		}()
	}
	go func() {
		wg.Wait()
		close(results)
	}()
	part2 := math.MaxInt
	for val := range results {
		if val < part2 {
			part2 = val
		}
	}
	// 2:54.71 without parallel
	// 55.6 sec with parallel
	fmt.Println(part2)
}

File: d5/main.py
text = open("d5/in.txt").read()
seed_block, *blocks = text.split("\n\n")
seeds = [int(s) for s in seed_block[7:].split()]
maps = [[[int(d) for d in l.split()] for l in b.strip().split("\n")[1:]] for b in blocks]

def F(seed: int) -> int:
    cur = seed
    for map_ in maps:
        for dest,source,rang in map_:
            if source <= cur < source + rang:
                cur = dest+cur-source
                break
    return cur

result = []
for seed in seeds:
    result.append(F(seed))
print(f"part 1: {min(result)}")

result = []
for start, count in zip(seeds[::2], seeds[1::2]):
    print(f"{count} interval...")
    for seed in range(start, start+count):
        result.append(F(seed)) 
    print(f"{count} done")
#part 2: 2520479
#pypy d5/main.py  3232.78s user 50.10s system 99% cpu 54:43.54 total
print(f"part 2: {min(result)}")
